lib/waive.py

```diff
@@ -0,0 +1,215 @@
+"""
+Functionality for an automatic failure waiving logic, configured by
+a custom file format. See WAIVES.md.
+"""
+
+import os
+import re
+import textwrap
+from pathlib import Path
+
+import util
+import versions
+
+_sections_cache = None
+
+WAIVE_FILE = Path(util.libdir).parent / 'conf' / 'waives'
+
+
+class _PushbackIterator:
+    """
+    A simple wrapper for an iterator that allows the user to "push back"
+    the last retrieved item, so it appears again in next().
+    """
+    def __init__(self, source):
+        self.it = iter(source)
+        self.pushed_back = False
+        self.last = None
+        self.counter = 0
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.pushed_back:
+            self.pushed_back = False
+            return self.last
+        else:
+            new = next(self.it)
+            self.last = new
+            self.counter += 1
+            return new
+
+    def pushback(self):
+        self.pushed_back = True
+
+
+class WaiveParseError(SyntaxError):
+    """
+    Easy waive file syntax error reporting, with line numbers derived
+    from _PushbackIterator style counter.
+    """
+    def __init__(self, meta, msg):
+        super().__init__(f"waive line {meta.counter}: {msg}")
+
+
+def _compile_eval(meta, code):
+    if not code.strip():
+        raise WaiveParseError(meta, "empty python block code ending here")
+    try:
+        return compile(textwrap.dedent(code), 'waivecode', 'eval')
+    except Exception:
+        raise WaiveParseError(meta, "compiling waive python code failed")
+
+
+def _parse_waive_file(stream):
+    sections = []
+    regexes = set()
+    python_code = ''
+    state = 'skipping_empty_lines'
+
+    lines = _PushbackIterator(stream)
+    for line in lines:
+        if line.startswith('#'):
+            continue
+        line = line.rstrip('\n')
+
+        if state == 'skipping_empty_lines':
+            if line:
+                state = 'reading_regex'
+                lines.pushback()
+                continue
```

Comment:
I'd prefer making the "empty skipping" more obvious there:
```suggestion
            if not line:
                continue
            else:
                state = 'reading_regex'
                lines.pushback()
```